File: 0907-koko-eating-bananas/0907-koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        
        int left = 1;
        int right = 0;

        for (int p : piles) {
            right = Math.max(right, p);
        }

        int k = right;
        while (left <= right) {
            int mid = left + (right - left) / 2;

            long hours = 0;
            for (int p : piles) {
                // same as ceil
                hours += (p + mid - 1) / mid;
            }

            if (hours <= h) {
                k = mid;
                right = mid - 1;
            }
            else left = mid + 1;
        }
        return k;
        
    }

}

File: 0907-koko-eating-bananas/0907-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        l, r = 1, max(piles)

        k = r

        while l <= r:
            mid = (l + r) // 2

            total = 0
            for p in piles:
                total += math.ceil(float(p) / mid)
            if total <= h:
                k = mid
                r = mid - 1
            else:
                l = mid + 1
        return k